staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go

```diff
@@ -0,0 +1,216 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dynamiccertificates
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	"k8s.io/client-go/util/cert"
+)
+
+func TestServingCert(t *testing.T) {
+	tlsConfig := &tls.Config{
+		// Can't use SSLv3 because of POODLE and BEAST
+		// Can't use TLSv1.0 because of POODLE and BEAST using CBC cipher
+		// Can't use TLSv1.1 because of RC4 cipher usage
+		MinVersion: tls.VersionTLS12,
+		// enable HTTP2 for go's 1.7 HTTP Server
+		NextProtos: []string{"h2", "http/1.1"},
+	}
+
+	defaultCertProvider, err := createTestTLSCerts(testCertSpec{
+		host:  "172.30.0.1",
+		ips:   []string{"172.30.0.1"},
+		names: []string{"kubernetes", "kubernetes.default.svc.cluster.local"},
+	}, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var sniCerts []SNICertKeyContentProvider
+	certSpecs := []testCertSpec{
+		{
+			host:  "172.30.0.1",
+			ips:   []string{"172.30.0.1"},
+			names: []string{"openshift", "openshift.default.svc.cluster.local"},
+		},
+		{
+			host:  "127.0.0.1",
+			ips:   []string{"127.0.0.1"},
+			names: []string{"localhost"},
+		},
+		{
+			host:  "2001:abcd:bcda::1",
+			ips:   []string{"2001:abcd:bcda::1"},
+			names: []string{"openshiftv6", "openshiftv6.default.svc.cluster.local"},
+		},
+	}
+
+	for _, certSpec := range certSpecs {
+		names := append([]string{}, certSpec.ips...)
+		names = append(names, certSpec.names...)
+		certProvider, err := createTestTLSCerts(certSpec, names)
+		if err != nil {
+			t.Fatal(err)
+		}
+		sniCerts = append(sniCerts, certProvider)
+	}
+
+	dynamicCertificateController := NewDynamicServingCertificateController(
+		*tlsConfig,
+		&nullCAContent{name: "client-ca"},
+		defaultCertProvider,
+		sniCerts,
+		nil, // TODO see how to plumb an event recorder down in here. For now this results in simply klog messages.
+	)
+	if err := dynamicCertificateController.RunOnce(); err != nil {
+		t.Fatal(err)
+	}
+	tlsConfig.GetConfigForClient = dynamicCertificateController.GetConfigForClient
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	server := &http.Server{
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			w.WriteHeader(http.StatusOK)
+		}),
+		MaxHeaderBytes: 1 << 20,
+		TLSConfig:      tlsConfig,
+	}
+	listener, _, err := createListener("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+	apiPort := listener.Addr().String()
+	go func() {
+		t.Logf("listening on %s", listener.Addr().String())
+		listener = tls.NewListener(listener, server.TLSConfig)
+		if err := server.Serve(listener); err != nil {
```

Comment:
does this actually start a TLS server? I expected .ServeTLS(...) (which also requires setting GetCertificate)

Reply:
nevermind, you already had a tls listener set up

Reply:
> does this actually start a TLS server? I expected .ServeTLS(...) (which also requires setting GetCertificate)

It uses the listener which is configured with TLS configuration. I'll update the line to make it more obvious.